pkg/probe/turbo_probe.go

```diff
@@ -9,36 +9,63 @@ import (
 )
 
 // Turbo Probe Abstraction
-// Consists of clients that handle probe registration metadata, discovery and action execution for different probe targets
+// Consists of clients that handle probe registration metadata,
+// and the discovery and action execution for different probe targets
 type TurboProbe struct {
-	ProbeType          string
-	ProbeCategory      string
-	RegistrationClient TurboRegistrationClient
-	DiscoveryClientMap map[string]TurboDiscoveryClient
+	ProbeConfiguration *ProbeConfig
+	RegistrationClient *ProbeRegistrationAgent
+	DiscoveryClientMap map[string]*TargetDiscoveryAgent
 
 	ActionClient TurboActionExecutorClient
 }
 
+type ProbeRegistrationAgent struct {
+	ISupplyChainProvider
+	IAccountDefinitionProvider
+	IActionPolicyProvider
+	IEntityMetadataProvider
+	discoveryMetadata *DiscoveryMetadata
+}
+
 type TurboRegistrationClient interface {
-	GetAccountDefinition() []*proto.AccountDefEntry
-	GetSupplyChainDefinition() []*proto.TemplateDTO
-	GetIdentifyingFields() string
-	// TODO: - add methods to get entity metadata, action policy data
+	ISupplyChainProvider
+	IAccountDefinitionProvider
+}
+
+func NewProbeRegistrator() *ProbeRegistrationAgent {
+	registrator := &ProbeRegistrationAgent{}
+	registrator.discoveryMetadata = NewDiscoveryMetadata()
+	return registrator
+}
+
+type TargetDiscoveryAgent struct {
```

Comment:
Shall we merge the FullDiscovery with the `IIncrementalDiscovery` and `IPerformanceDiscovery`?
then the probe writer should explicitly implement them all, while the later two can be empty functions?